pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -466,26 +466,26 @@ func (s *DiscoveryServer) pushEds(push *model.PushContext, con *XdsConnection, v
 // getDestinationRule gets the DestinationRule for a given hostname. As an optimization, this also gets the service port,
 // which is needed to access the traffic policy from the destination rule.
 func getDestinationRule(push *model.PushContext, proxy *model.Proxy, hostname host.Name, clusterPort int) (*networkingapi.DestinationRule, *model.Port) {
-	for _, service := range push.Services(proxy) {
-		if service.Hostname == hostname {
-			cfg := push.DestinationRule(proxy, service)
-			if cfg == nil {
-				continue
-			}
-			for _, p := range service.Ports {
-				if p.Port == clusterPort {
-					return cfg.Spec.(*networkingapi.DestinationRule), p
-				}
-			}
+	cfg := push.DestinationRule(proxy, &model.Service{Hostname: hostname})
```

Comment:
@hzxuzhonghu Does this change make sense to you? I am trying to optimize instead of looping through all push services - directly pick using a dummy service - which covers most of the cases.

Also do we need to get the service and match port below? or can we just use &model.Port{clusterPort} and match against the port level settings in destination rule? If yes, we can avoid that lookup as well - I am not able to think of a case - where that does not work?

Reply:
I am concerned with this case: we have two SE resides in two different namespaces, we may not get the related DR correctly.

Reply:
Yeah, I thought about that, But does n't the current implementation also do the same? it just picks the service that matches host name first?
https://github.com/istio/istio/blob/master/pilot/pkg/model/push_context.go#L764

Reply:
Yeah, just saw that it is not quite correct. For the gateway type, it jumps down to the below logic.

Reply:
What should we do for this then? Should I add this 
```
proxy.SidecarScope != nil && proxy.Type == SidecarProxy && len(ps.ServiceByHostnameAndNamespace[service.Hostname]) == 1

Reply:
With a close look at the dr selection orders, i think this is not a matter.

Reply:
OK. Are you saying what this PR has is fine?

Also can you confirm if this can be done?

Do we need to get the service and match port below? or can we just use &model.Port{clusterPort} and match against the port level settings in destination rule? If yes, we can avoid that lookup as well - I am not able to think of a case - where that does not work?

Reply:
@ramaraochavali To totally solve this issue, you need to change `SidecarScope.destinationRules` type to map[host.Name]map[namespace]*Config

Reply:
Yeah. That is what I thought too. For now, I am thinking of changing what is there in this PR - at least it improves perf with the current behaviour. WDYT?

Reply:
make sense

Reply:
Ok. Changed PTAL